src/algebra/associated.lean

```diff
@@ -547,6 +550,56 @@ begin
   rw [mk_mul_mk, mk_dvd_mk, mk_dvd_mk, mk_dvd_mk],
 end
 
+theorem irreducible_mk_iff (a : α) : irreducible (associates.mk a) ↔ irreducible a :=
+begin
+  simp [irreducible, is_unit_mk],
+  apply and_congr iff.rfl,
+  split,
+  { assume h x y eq,
+    have : is_unit (associates.mk x) ∨ is_unit (associates.mk y),
+      from h _ _ (by rw [eq]; refl),
+    simpa [is_unit_mk] },
+  { refine assume h x y, quotient.induction_on₂ x y (assume x y eq, _),
+    rcases quotient.exact eq.symm with ⟨u, eq⟩,
+    have : a = x * (y * u), by rwa [mul_assoc, eq_comm] at eq,
+    show is_unit (associates.mk x) ∨ is_unit (associates.mk y),
+    simpa [is_unit_mk] using h _ _ this }
+end
+
+theorem mk_dvd_not_unit_mk_iff {a b : α} :
+  (associates.mk a ≠ 0 ∧ ∃ x, ¬is_unit x ∧ associates.mk b = associates.mk a * x) ↔
+  a ≠ 0 ∧ ∃ x, ¬is_unit x ∧ b = a * x :=
```

Comment:
The `dvd_not_unit` part of this theorem pops up quite a few times. It might be worth defining `def dvd_not_unit {a b : α} : Prop := a ≠ 0 ∧ ∃ x, ¬is_unit x ∧ b = a * x`. It is equivalent (with the correct assumptions) to `a | b ∧ ¬ b | a`, i.e. `(<) : associates α -> associates α -> associates α`, right?

Reply:
Yeah. it's equivalent to `a | b ∧ ¬ b | a` if you have cancellation, and one of the implications is always true.
I thought about making the core definition of `DCC_dvd` well-foundedness of `<` on `associates`, but then decided to use this relation because of the theorems that already existed for Noetherian rings.